polars/polars-io/src/csv/write_impl.rs

```diff
@@ -44,30 +43,16 @@ fn write_anyvalue(f: &mut Vec<u8>, value: AnyValue, options: &SerializeOptions)
         AnyValue::UInt16(v) => write!(f, "{}", v),
         AnyValue::UInt32(v) => write!(f, "{}", v),
         AnyValue::UInt64(v) => write!(f, "{}", v),
-        AnyValue::Float32(v) => {
-            let len = f.len();
-            let write_size = f32::FORMATTED_SIZE_DECIMAL;
-            f.reserve(write_size);
-            unsafe {
-                let buf = std::slice::from_raw_parts_mut(f.as_mut_ptr().add(len), write_size);
-
-                let written_n = v.to_lexical(buf).len();
-                f.set_len(len + written_n);
-            }
-            Ok(())
-        }
-        AnyValue::Float64(v) => {
-            let len = f.len();
-            let write_size = f64::FORMATTED_SIZE_DECIMAL;
-            f.reserve(write_size);
-            unsafe {
-                let buf = std::slice::from_raw_parts_mut(f.as_mut_ptr().add(len), write_size);
-
-                let written_n = v.to_lexical(buf).len();
-                f.set_len(len + written_n);
-            }
-            Ok(())
-        }
+        AnyValue::Float32(v) => match &options.float_precision {
```

Comment:
This code was heavily optimized. We now do a redundant heap allocation for every float we write. This will be a regression.

We need to write directly to the buffer. Can we adapt the old code to control the precision?

Reply:
Oh, interesting — I didn’t realize! I will instead try to adapt the old code. 

For my own benefit, could you elaborate on how the old code avoids a double heap allocation? Is the problem in ‘write!’?

Reply:
The `write!` macro accepts a formatter/buffer to which it will write. In our case that is `f: &mut Vec<u8>` which is our intermediate buffer. This intermediate buffer will be flushed to disk when the thread has finished its batch.

With the new changes you first call `to_string` which converts the float to a string value on the heap.